pythonfuzz/fuzzer.py

```diff
@@ -134,9 +156,9 @@ def start(self):
                 break
 
             buf = self._corpus.generate_input()
-            parent_conn.send_bytes(buf)
+            parent_conn.send_bytes(bytes(buf))
```

Comment:
Isn't it bytes already? I think this might slow down the fuzzer if it's an unnecessary copy.

Reply:
There's a significant difference in how Python 2 and Python 3 handle 'bytes'.
In python 3, `bytes` is a list of vaules which when iterated over, or indexed, are just integers.
In python 2, `bytes` is a str, which when interated over, or indexed, contains single character `str`.

This difference means that working in the mutators with bytes is impractical - the code would need to handle both the cases of the elements of the iteration being integers and being str objects. That would be... tiresome and difficult to work with... but, if instead the elements are handled as bytearrays, this situation is simplified.

In both python 2 and python 3, bytearray is able to be initialised just like a bytes object, and iterates as a set of integers. It can be converted to a bytes object easily without encoding issues, and manipulated just like we were handling the bytes object. Because we're actually working with a list of bytes which we want to mutate, a bytearray is actually the ideal container to use. The python 3 documentation describes it as:

> The bytearray class is a mutable sequence of integers in the range 0 <= x < 256. It has most of the usual methods of mutable sequences, described in Mutable Sequence Types, as well as most methods that the bytes type has...

(https://docs.python.org/3/library/functions.html#func-bytearray)

I originally moved to bytearray because 'it just worked', but as I looked at what it was, it became more obvious that it's actually the right tool for the job. There is, obviously, the fact that the conversion may incur a little cost, but compared to the other operations and the transfers between processes that are performed, I suspect that it's entirely negligable.